### PR TITLE
DateAdd and DateSubtract expression

### PR DESCRIPTION
**Pull request by qnkhuat**
Implement `:date-add` and `:date-subtract` methods as spec'ed in this [project doc](https://www.notion.so/metabase/Unlock-advanced-date-time-manipulation-038f5cf80f364d7cb93341784169743c)

date-add
Screen Shot 2022-09-20 at 23 16 58

date-subtract
Screen Shot 2022-09-20 at 23 17 13

### CODE REVIEW

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (metamben) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:181]: Use `(testing ...)` to name the test cases.

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (metamben) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:185]: Collapse these into a single `doseq` and save some horizontal space.

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (metamben) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:155]: Don't we want to eliminate this difference? (I could argue pro and contra, so I'm just wondering.)

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (metamben) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:187]: I wonder if it was easier to follow if the field values were read from the database instead of being hardwired here.

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (metamben) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:271]: You could `doseq` these two as well.

**Review by metamben — COMMENTED** [review]
Looks reasonable, but I can't really judge the completeness. I hope someone else will have a look at it too.

File: shared/src/metabase/mbql/schema.cljc
Comment (qnkhuat) [reply shared/src/metabase/mbql/schema.cljc:634]: ah right, should be ExpressionArg to match other extract functions.

While we're on this, do you think we need a whole expression for datetime?

**Review by ranquild — APPROVED** [review]
FE 👍

File: modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj
Comment (dpsutton) [inline modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj:188]: I'm surprised this is necessary. When setting a corecion_strategy the effective_type should be set to the resulting type. (eg, base type is string, coercion is Coercion/ISO8601->DateTime, and effective is :type/DateTime or equivalent). The fact that this needed to be done makes me think we have some bug elsewhere if this is not happening.

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (dpsutton) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:174]: I think this should be made resilient to the version string not being present. The `str/split` and `parse-long` will both error out if that version string is missing. If we do want to error we should have a better error message than those will provide.

File: modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj
Comment (qnkhuat) [reply modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj:188]: > The fact that this needed to be done makes me think we have some bug elsewhere if this is not happening.

You're right, and the bug is the mapping between our type system and the actual type we cast is mismatched for bigquery.

Specifically when you do a `:Coercion/ISO8601->DateTime` we do 2 things:
- we set the effective type of this column to `type/DateTime`
- we cast this column to `timestamp` ([ref](https://github.com/metabase/metabase/blob/a99ec1dfbb5bccaed4c827054df738789a0a0ea1/modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj#L728))

The mismatch here is we interpret `type/Datetime` as `datetime` type, but the actually db-type is `timestamp` (because we cast to timestamp).

This is a problem now because bigquery distinguishes between `timestamp` and `datetime` when you do datetime arithmetic.
See [this](https://github.com/metabase/metabase/blob/8fa5b3c2e09aa9531058ff940e1ee019f63ea527/modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj#L721) comment for details. 

--- 

I think the correct fix here is to make sure the effective type for `:Coercion/ISO8601->DateTime` is `type/DateTimeWithTZ`, this way we will interpret it correctly as `timestamp`.

But the [code](https://github.com/metabase/metabase/blob/8fa5b3c2e09aa9531058ff940e1ee019f63ea527/shared/src/metabase/types.cljc#L355) for this is not dispatched per driver, And considering how a big `radius` this change affect our codebase, I think it's better to solve it within bigquery like I did here.

WDYT?

Comment (qnkhuat) [reply modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj:188]: actually, we could just make `:Coercion/ISO8601->DateTime` cast to `datetime` for bigquery.
Any reason this is a no-go?

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (qnkhuat) [reply test/metabase/query_processor_test/date_time_zone_functions_test.clj]: https://github.com/metabase/metabase/pull/25414/commits/9d32c7b56af1f55b292678d582d0e3183f7078f4

I make the behavior of vertica consistent with other dbs in this commit. WDYT?

File: modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj
Comment (qnkhuat) [reply modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj]: I went ahead with making `:Coercion/ISO8601->DateTime` to cast to `datetime` in this commit https://github.com/metabase/metabase/pull/25414/commits/786970c24f11c2ea8fe5277e42b9fa80b23a798e

**Comment by qnkhuat** [comment]
Hi @nllho. could you look into the docs for these 2 news `date-add` and `date-subtract` expressions? I have screenshots of it in the PR's description.

**Comment by nllho** [comment]
@qnkhuat Here are my suggestions :)

I changed "datetime" to "date or timestamp value" because `datetime` is a specific data type in some databases.

But if the function _only_ works with specific data types (`datetime`, `timestamp`, `date`, etc.), then feel free to replace "date or timestamp value" with whatever is correct!

```
dateAdd(column, amount, unit)

Adds some amount of time to a date or timestamp value.

column

The name of the column with your date or timestamp values.

amount

The number of time units to be added.

unit

"year", "month", "quarter", "day", "hour", "minute", or "second".

```

```
dateSubtract(column, amount, unit)

Subtracts some amount of time from a date or timestamp value.

column

The name of the column with your date or timestamp values.

amount

The number of time units to be subtracted.

unit

"year", "month", "quarter", "day", "hour", "minute", or "second".
```

**Comment by qnkhuat** [comment]
@nllho Thanks for look into this. I have some feedback.
> Adds some amount of time to a date or timestamp value.

I'm concerned about the `amount of time` part, this _might_ suggest that this only work with time(hour, minute, second) but it actually work with date  unit too (day, month ,quarter, year)

> The name of the column with your date or timestamp values.

I'd avoid using `the name of the column`  because it's technically incomplete. The `column` could be the result of an expression that returns a datetime column. Think case where we do `dateAdd(convertTimezone([created at], "Asia/Tokyo"), -1, "month")`

FWIW my version took some inspiration from this doc : https://www.techonthenet.com/excel/formulas/dateadd.php

**Comment by nllho** [comment]
@qnkhuat Good points! I think we just want to make it clear that "units" refers to the "unit" parameter, which tells people the size of the time units that they can choose from.

Here's an updated suggestion :)

```
dateAdd(column, amount, unit)

Adds some units of time to a date or timestamp value.

column

The column with your date or timestamp values.

amount

The number of units to be added.

unit

"year", "month", "quarter", "day", "hour", "minute", or "second".
```

**Review by metamben — APPROVED** [review]
LGTM

**Comment by qnkhuat** [comment]
Updated! Thank you @nllho

File: shared/src/metabase/mbql/schema.cljc
Comment (camsaul) [inline shared/src/metabase/mbql/schema.cljc:651]: 1. Should `:millisecond` be in here as well? Since some of the code above is written to support it.
2. I think we should disallow `:default` here. How does one add a `1 :default`? It doesn't make sense. We should require an explicit unit. When we use `:default` in other places it basically means "use this column at it's current resolution, don't apply truncation/bucketing"... doesn't really apply here.

```suggestion
   (apply s/enum #{:millisecond :second :minute :hour :day :week :month :quarter :year})
```

File: shared/src/metabase/mbql/schema.cljc
Comment (camsaul) [inline shared/src/metabase/mbql/schema.cljc:655]:
```suggestion
  datetime DateTimeExpressionArg
```

File: shared/src/metabase/mbql/schema.cljc
Comment (camsaul) [inline shared/src/metabase/mbql/schema.cljc:660]:
```suggestion
  datetime DateTimeExpressionArg
```

File: shared/src/metabase/mbql/schema.cljc
Comment (camsaul) [inline shared/src/metabase/mbql/schema.cljc:865]: I'd personally rather not leave a bunch of comments for testing stuff all over the codebase. Take this and write an actual test

```suggestion
```

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (camsaul) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:14]: Can you give this a better name? `(formatting x)` is meaningless. `(format-timestamp x)` is much better, but you can probably do something better than that. Maybe `normalize-timestamp-str` or something.

I would also not use this for anything but timestamp strings. You should know which columns are which in the tests... use `int` for the ID column and `normalize-timestamp-str` for the timestamp columns.

```suggestion
(defn- normalize-timestamp-str [x]
  (-> x
      (str/replace  #"T" " ")
      (str/replace  #"Z" "")))
```

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (camsaul) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:271]: Mongo-specific tests belong somewhere `modules/drivers/mongo/test`, not in the general tests. Make some of the stuff here public so you can reuse it there if you need to.

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (camsaul) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:206]: Suggestions:

Could you make column type part of the `doseq` as well? That way you have one mega `doseq` instead of three

These sorts of big `doseq`-based tests get sort of hard to read I think. Since a lot of the stuff below is declarative I'd consider moving what you can out into a top-level `def` and call it `date-math-test-cases` or something. Then the actual business logic can go in the test itself. It makes things a little easier on the eyes I think.

Just remember that you shouldn't be loading the dataset or calling `mt/id` in a top-level form... so your `def` should have column name (`:dt`) and the test itself can use that to get the actual Field

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (camsaul) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:299]: If you're excluding a driver you should leave a comment with a note as to why you are doing so. If it's because there is a different version of the test for mongo then include a `[[link]]` to it in your comment

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (camsaul) [reply test/metabase/query_processor_test/date_time_zone_functions_test.clj:14]: You probably don't even want to use this in `test-date-extract` anyway, right? I don't understand why you changed **that** test

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (camsaul) [reply test/metabase/query_processor_test/date_time_zone_functions_test.clj:197]: I don't really love the pattern of checking the MongoDB version here. If supporting this stuff is dependent on the Mongo version, we should be using `driver/database-supports?` to check support for the feature rather than `driver/supports?`. That being said we don't have test tooling around that yet -- `mt/normal-drivers-with-feature` checks `driver/supports?` rather than `driver/database-supports?`. So I won't put my foot down on this.

That being said if you want to do this the right way you can always do something like

```clj
(mt/test-drivers (mt/normal-drivers-with-feature :date-arithmetics)
  (when (driver/database-supports? driver/*driver* (mt/db) :date-arithmetics)
    ...))
```

Better yet, you can turn that into a test util macro so we can start using that everywhere.

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (camsaul) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:204]: you gotta write a note about why we're skipping Mongo here.

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (camsaul) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:213]: I know tuples like `[expected query]` are more convenient than maps but seeing `[[[[` makes my eyes bleed... I think 

```clj
{:expected ...
 :query ...}
```

would be a bit more readable. Maybe if you move this stuff into a top-level `def` you can do that instead.

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (camsaul) [reply test/metabase/query_processor_test/date_time_zone_functions_test.clj:206]: For the query you could change it to be a function maybe e.g. `:query-fn (fn [field-id] ...)`. IDK. Or just have it look at a dynamic var. Doesn't really matter. I just want this test to be easier to read

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (camsaul) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:196]:
```suggestion
(defn- db-major-version [db]
```

Is this Mongo DB specific at all?

**Review by camsaul — APPROVED** [review]
Overall it looks good to me but I left a handful of notes about things. To summarize my suggestions:

- [x] Should we explicitly support `:millisecond`? I would vote yes just for completeness.
- [x] I think adding or subtracting a `:default` makes no sense. You should probably take that out
- [x] If you skip tests for a specific driver you have to put a comment in there explaining why. If it's because there's a different version of the test somewhere else then include a `[[link]]` to it in the comments. Ideally, you wouldn't skip tests for Mongo at all... maybe if you do the `driver/database-supports?` thing I suggested you wouldn't need to skip it for Mongo
- [x] Driver-specific tests should go in driver test namespaces. Ideally there should be zero occurrences of the `:mongo` keyword in `test/` at all. I know that's not the reality but it's something we should stive for
- [x] **Tests** should not be checking database version... that's a clear sign we need to implement `driver/database-supports?` rather than `driver/supports?`. Please implement it for MongoDB and have that method check the major version, and then update your test to call that as I suggested
- [x] The main test is very big and a little hard to read. See if there's a way to simplify it somehow, maybe by taking stuff like `:expected`/`:query` out and putting it in a top-level `def`.

Since all of those things are relatively minor I'm going to conditionally approve this PR so we you don't need to wait for another review before getting this in. So once you address my feedback feel free to merge.

File: shared/src/metabase/mbql/schema.cljc
Comment (qnkhuat) [reply shared/src/metabase/mbql/schema.cljc:865]: Oh I forgot to delete these from debugging. Will move.

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj
Comment (qnkhuat) [reply test/metabase/query_processor_test/date_time_zone_functions_test.clj:233]: If we want to move this to a top-level def, then the expected need to be an `fn` too, because the `expected` are being generated dynamically.

So it'll probably look something like

```clojure
{:expected-fn (fn [col-type op unit col-type]... )
 :query-fn    (fn [field-id] ...))
```
I'm not sure this makes it easier to reason about the tests tho.

But I'm all ears on how to improve these tests, I've been going back and forth with it several times and still not fully satisfied with it.

**Pull request merged**
